### PR TITLE
Enhance reduce and transpose op

### PR DESCRIPTION
**Pull request by joey12300**
### PR types

Function optimization
### PR changes

OPs
### Describe

* enhance reduce op which can reduce tensor with arbitrary rank (rank >= 7).
* enhance transpose op which can transpose tensor with arbitrary rank (rank>=7).
* add eigen device thread pool to accelerate transpose function.
* update transpose, reduce unittest to coverage situation when tensor has rank >= 7

### CODE REVIEW

File: paddle/fluid/operators/math/math_function.cc [hunk 1]
```diff
@@ -63,6 +65,55 @@ DEFINE_CPU_TRANS(4);
...
+template <typename T>
+struct TransposeNormal<platform::CPUDeviceContext, T> {
+  void operator()(const platform::CPUDeviceContext& context,
+                  const framework::Tensor& in, framework::Tensor* out,
+                  const std::vector<int>& axis) {
+    const int rank = axis.size();
+    auto in_stride = framework::stride(in.dims());
+    auto out_stride = framework::stride(out->dims());
+    const T* in_ptr = reinterpret_cast<const T*>(in.data<T>());
+    T* out_ptr = reinterpret_cast<T*>(out->data<T>());
```
Comment (wawltor) [inline paddle/fluid/operators/math/math_function.cc]: 这里看起来不需要使用reinterpret_cast，后面已经使用了template

File: paddle/fluid/operators/reduce_ops/reduce_op.h [hunk 2]
```diff
@@ -34,6 +35,110 @@ namespace operators {
...
+  // transpose to 2D tensor whose shape is {unreduced, reduced}.
+  const int64_t unreduced = output->numel();
+  const int64_t reduced = shuffled_input.numel() / unreduced;
+  shuffled_input.Resize({unreduced, reduced});
+  DDim output_dim = output->dims();
+  output->Resize({unreduced});
+  ReduceFunctor<DeviceContext, OutT, 2, 1, Functor>(
+      context.template device_context<DeviceContext>(), shuffled_input, output,
+      {1}, keep_dim);
+  output->Resize(output_dim);
```
Comment (wawltor) [inline paddle/fluid/operators/reduce_ops/reduce_op.h:108]: 这里的outpu_dim是否要考虑keep_dim的情况了

File: paddle/fluid/operators/math/math_function.cu [hunk 3]
```diff
@@ -46,6 +51,88 @@ DEFINE_GPU_TRANS(4);
...
+      cpu_buf[rank + i] = out_stride[i];
+      cpu_buf[2 * rank + i] = axis[i];
+    }
+    memory::Copy(cuda_place, cuda_buf, cpu_place, cpu_buf, size,
+                 context.stream());
+    REINTERPRET(const int64_t, in_stride_ptr, cuda_buf);
+    REINTERPRET(const int64_t, out_stride_ptr, cuda_buf + rank);
+    REINTERPRET(const int64_t, axis_ptr, cuda_buf + 2 * rank);
+
+    constexpr int MAX_BLOCK_DIM = 512;
```
Comment (wawltor) [inline paddle/fluid/operators/math/math_function.cu]: 不同的cpu卡 block里面的thread_num不同，用GetMaxThreadsPerBlock代替

File: paddle/fluid/operators/math/math_function.cu [hunk 4]
```diff
@@ -46,6 +51,88 @@ DEFINE_GPU_TRANS(4);
...
+template <typename T>
+struct TransposeNormal<platform::CUDADeviceContext, T> {
+  void operator()(const platform::CUDADeviceContext& context,
+                  const framework::Tensor& in, framework::Tensor* out,
+                  const std::vector<int>& axis) {
+    const int rank = axis.size();
+    auto in_stride = framework::stride(in.dims());
+    auto out_stride = framework::stride(out->dims());
+    REINTERPRET(const T, in_ptr, in.data<T>());
+    REINTERPRET(T, out_ptr, out->data<T>());
```
Comment (wawltor) [inline paddle/fluid/operators/math/math_function.cu]: 这里看起来不需要cast

File: paddle/fluid/operators/math/math_function.cc (hunk 1 above)
Comment (joey12300) [reply paddle/fluid/operators/math/math_function.cc]: done

File: paddle/fluid/operators/math/math_function.cu (hunk 3 above)
Comment (joey12300) [reply paddle/fluid/operators/math/math_function.cu]: done

File: paddle/fluid/operators/math/math_function.cu (hunk 4 above)
Comment (joey12300) [reply paddle/fluid/operators/math/math_function.cu]: @wawltor done

**Review by wawltor — APPROVED** [review]
LGTM

**Pull request merged**